### PR TITLE
Support collection deletion in apiserver.

### PR DESCRIPTION
**Pull request by wojtek-t**
What does it give us:
- in scalability tests, namespace deletion takes ~3m instead of ~30m
- the API is pretty generic, although the first implementation is not the ideal one - this should work much better in etcd v3 api

@derekwaynecarr @smarterclayton @lavalamp @gmarek 

Ref #7372

### CODE REVIEW

**Comment by wojtek-t** [comment]
@smarterclayton @derekwaynecarr - can you please take a first look? If you're fine with the approach, I will extend test, etc.

File: pkg/apiserver/resthandler.go [hunk 1]
```diff
@@ -703,6 +703,99 @@ func DeleteResource(r rest.GracefulDeleter, checkBody bool, scope RequestScope,
...
+			return
+		}
+
+		ctx := scope.ContextFunc(req)
+		ctx = api.WithNamespace(ctx, namespace)
+
+		if admit != nil && admit.Handles(admission.Delete) {
+			userInfo, _ := api.UserFrom(ctx)
+
+			err = admit.Admit(admission.NewAttributesRecord(nil, scope.Kind.Kind, namespace, "", scope.Resource.Resource, scope.Subresource, admission.Delete, userInfo))
```
Comment (derekwaynecarr) [inline pkg/apiserver/resthandler.go]: I am wondering if admission control should make it clearer when you are doing a delete over an entire collection rather than just DELETE + kind + (optional namespace).  It's probably fine as is, but will mull on it some.

File: pkg/registry/generic/etcd/etcd.go [hunk 2]
```diff
@@ -440,6 +441,33 @@ func (e *Etcd) Delete(ctx api.Context, name string, options *api.DeleteOptions)
 	return e.finalizeDelete(out, true)
 }
 
+func (e *Etcd) DeleteCollection(ctx api.Context, options *api.DeleteOptions, listOptions *unversioned.ListOptions) (runtime.Object, error) {
```
Comment (derekwaynecarr) [inline pkg/registry/generic/etcd/etcd.go]: I am wondering if the response should be the following:

1. all the items that were in the listing scope (as i see it is now)
2. all the items that were actually deleted
3. all the items that did not delete 

Since we return immediately on error, I guess (3) is not really possible, and if that is the case, (2) and (1) become the same result.

Comment (derekwaynecarr) [inline pkg/registry/generic/etcd/etcd.go]: @wojtek-t - does the etcd v3 api remove the concept of a directory node?  if we only intend to use this from the context of namespace deletion (where we know creates are blocked), then it would be good if after all the delete nodes, we delete the immediate parent directory node in etc.

**Comment by derekwaynecarr** [comment]
I am fine with the approach.

File: pkg/registry/generic/etcd/etcd.go (hunk 2 above)
Comment (wojtek-t) [inline pkg/registry/generic/etcd/etcd.go]: Regarding the first one - yes - it's exactly as you wrote. I think it's the cleanest semantics, although we probably need to be clear that the operation is (at least for now) NOT atomic (i.e. it can happen that some of the elements will be removed, before error is returned.

Regarding the second comment - AFAIK yes - they are removing the concept of directory.
I agree that ideally we should remove also the "directory" entry in etcd, but currently it's not doable with our API. I will extend this TODO to be clear about.
Also note, that this PR is NOT making things worse (in fact we are moving "bulk deletion" from client to server. The result in the etcd is exactly the same.

File: pkg/apiserver/resthandler.go (hunk 1 above)
Comment (wojtek-t) [inline pkg/apiserver/resthandler.go]: I'm far from being an expert in admission. I did something that was the easiest to do, but if you think we should do something different, I'm happy to change it.

**Comment by wojtek-t** [comment]
@derekwaynecarr - thanks!

I will cleanup this PR, add some tests, etc. and will let you know (probably tomorrow).

File: pkg/registry/generic/etcd/etcd.go (hunk 2 above)
Comment (derekwaynecarr) [inline pkg/registry/generic/etcd/etcd.go]: make sense.

File: pkg/apiserver/resthandler.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/apiserver/resthandler.go]: This would require (for instance) admission controllers that track resource cluster usage to also track bulk deletion and implement the same logic that the list call does.  Generally we want admission controllers to operate on concrete objects.  At a minimum the list Options have to be passed in.

Comment (wojtek-t) [inline pkg/apiserver/resthandler.go]: @smarterclayton - not sure if I understood. Why do you want to pass ListOptions? Do you want to have it based on selectors?
Does it have to be part of this PR (I have almost zero knowledge about admission control).

File: pkg/api/rest/rest.go [hunk 3]
```diff
@@ -125,6 +125,13 @@ func (w GracefulDeleteAdapter) Delete(ctx api.Context, name string, options *api
 	return w.Deleter.Delete(ctx, name)
 }
 
+// CollectionDeleter is an object that can delete a collection
+// of RESTful resources.
+type CollectionDeleter interface {
```
Comment (timothysc) [inline pkg/api/rest/rest.go]: could we mark as TODO for future cleanup.

File: pkg/client/unversioned/events.go [hunk 4]
```diff
@@ -184,6 +185,30 @@ func (e *events) Delete(name string) error {
 		Error()
 }
 
+// DeleteCollections deletes a collection of objects.
+func (e *events) DeleteCollection(options *api.DeleteOptions, listOptions unversioned.ListOptions) error {
```
Comment (timothysc) [inline pkg/client/unversioned/events.go]: Maybe I missed it but what prevents multiple overlapping deletes of the same collection?

File: pkg/registry/generic/etcd/etcd.go [hunk 5]
```diff
@@ -440,6 +441,33 @@ func (e *Etcd) Delete(ctx api.Context, name string, options *api.DeleteOptions)
...
+		return nil, err
+	}
+	for _, item := range items {
+		accessor, err := meta.Accessor(item)
+		if err != nil {
+			return nil, err
+		}
+		// TODO: Shouldn't we copy options here?
+		// TODO: Shouldn't we read returned object?
+		if _, err := e.Delete(ctx, accessor.Name(), options); err != nil {
```
Comment (timothysc) [inline pkg/registry/generic/etcd/etcd.go]: you "could" even parallelize this operation... Dunno if you want to.  Namespace deletes are so painful in e2es.

File: pkg/api/rest/rest.go (hunk 3 above)
Comment (wojtek-t) [inline pkg/api/rest/rest.go]: Sorry - I don't understand this comment. What you would like to cleanup with it?

File: pkg/client/unversioned/events.go (hunk 4 above)
Comment (wojtek-t) [inline pkg/client/unversioned/events.go]: Actually nothing. But I don't think it's a problem - if the second one will try to delete object that is already deleted, it will return an error.

File: pkg/registry/generic/etcd/etcd.go (hunk 5 above)
Comment (wojtek-t) [inline pkg/registry/generic/etcd/etcd.go]: I agree - but even without any paralellization it's 3m instead of 35m in scalability tests.
I think it's good enough for the first step - we can do further improvements later.

**Comment by wojtek-t** [comment]
@derekwaynecarr @smarterclayton - I cleaned up this PR and it is ready for review
The e2e failures are not related to this PR - we have huge problems with our e2e now and are trying to fix them - will rerun once the problem is solved.

PTAL

File: pkg/api/rest/rest.go (hunk 3 above)
Comment (timothysc) [inline pkg/api/rest/rest.go]: Aren't we planning on changing this post etcd3?

Comment (wojtek-t) [inline pkg/api/rest/rest.go]: I don't think so. I think the interface should remain as it is here.

What we would like to cleanup is the implementation of DeleteCollection in pkg/registry - but there is already a TODO there.

File: pkg/registry/generic/etcd/etcd.go (hunk 5 above)
Comment (timothysc) [inline pkg/registry/generic/etcd/etcd.go]: That is great! @rrati will be thrilled.

File: pkg/api/rest/rest.go (hunk 3 above)
Comment (timothysc) [inline pkg/api/rest/rest.go]: sgtm

Comment (derekwaynecarr) [inline pkg/api/rest/rest.go]: Will review first thing tomorrow morning.

Comment (wojtek-t) [inline pkg/api/rest/rest.go]: @derekwaynecarr - great thanks!

**Comment by wojtek-t** [comment]
@derekwaynecarr - PTAL

File: pkg/api/rest/rest.go [hunk 6]
```diff
@@ -125,6 +125,13 @@ func (w GracefulDeleteAdapter) Delete(ctx api.Context, name string, options *api
...
+	// TODO: Add comment.
```
Comment (derekwaynecarr) [inline pkg/api/rest/rest.go]: Fix the // TODO and actually add the comment on how this is used, and when to use it.

File: pkg/apiserver/api_installer.go [hunk 7]
```diff
@@ -567,6 +571,24 @@ func (a *APIInstaller) registerResourceHandlers(path string, storage rest.Storag
 			}
 			addParams(route, action.Params)
 			ws.Route(route)
+		case "DELETECOLLECTION":
+			doc := "delete collection of " + kind
+			if hasSubresource {
+				doc = "delete collection of " + subresource + " of a " + kind
```
Comment (derekwaynecarr) [inline pkg/apiserver/api_installer.go]: not obvious to me, but this doesn't appear to be used.  where would this be useful?  am i missing an obvious kind?

File: api/swagger-spec/v1.json [hunk 8]
```diff
@@ -2826,6 +3126,75 @@
       "consumes": [
        "*/*"
       ]
+     },
+     {
+      "type": "unversioned.Status",
+      "method": "DELETE",
+      "summary": "delete collection of Node",
+      "nickname": "deletecollectionNamespacedNode",
```
Comment (derekwaynecarr) [inline api/swagger-spec/v1.json]: noting so I can see if I am not mistaken, but a node is not namespaced, is this a bug?

File: api/swagger-spec/v1.json [hunk 9]
```diff
@@ -4374,18 +4820,12 @@
...
+     },
      {
-      "type": "json.WatchEvent",
-      "method": "GET",
-      "summary": "watch individual changes to a list of PersistentVolume",
-      "nickname": "watchNamespacedPersistentVolumeList",
+      "type": "unversioned.Status",
+      "method": "DELETE",
+      "summary": "delete collection of PersistentVolume",
+      "nickname": "deletecollectionNamespacedPersistentVolume",
```
Comment (derekwaynecarr) [inline api/swagger-spec/v1.json]: same here, a PV is global I thought?  Only a PVC is namespaced if I recall.  Why is this saying namespaced

File: pkg/client/unversioned/events.go [hunk 10]
```diff
@@ -42,6 +42,7 @@ type EventInterface interface {
 	// Search finds events about the specified object
 	Search(objOrRef runtime.Object) (*api.EventList, error)
 	Delete(name string) error
+	DeleteCollection(options *api.DeleteOptions, listOptions unversioned.ListOptions) error
```
Comment (derekwaynecarr) [inline pkg/client/unversioned/events.go]: nit Godoc please

File: pkg/client/unversioned/events.go [hunk 11]
```diff
@@ -184,6 +185,30 @@ func (e *events) Delete(name string) error {
 		Error()
 }
 
+// DeleteCollections deletes a collection of objects.
```
Comment (derekwaynecarr) [inline pkg/client/unversioned/events.go]: DeleteCollection deletes

**Comment by derekwaynecarr** [comment]
@wojtek-t - so you are only looking to make the namespace controller use this for Events in this PR?  want to make sure I get the initial scope, or if you plan to extend this for other resource types?

**Comment by derekwaynecarr** [comment]
@deads2k - just a heads up, but this PR would allow for `DELETE /api/v1/namespace/foo/pods/` to do a DELETE_ALL, which I think is fine for our policy concerns, but wanted to give you a heads up.

File: pkg/api/rest/rest.go (hunk 6 above)
Comment (wojtek-t) [inline pkg/api/rest/rest.go]: done.

File: pkg/apiserver/api_installer.go (hunk 7 above)
Comment (wojtek-t) [inline pkg/apiserver/api_installer.go]: We have the "hasSubresource" code in pure "DELETE" path:
https://github.com/kubernetes/kubernetes/blob/master/pkg/apiserver/api_installer.go#L554

Since DeleteCollection is logicall a "collection of deletions", I think it makes sense to have it here.

File: pkg/client/unversioned/events.go (hunk 10 above)
Comment (wojtek-t) [inline pkg/client/unversioned/events.go]: done

File: pkg/client/unversioned/events.go (hunk 11 above)
Comment (wojtek-t) [inline pkg/client/unversioned/events.go]: done

File: api/swagger-spec/v1.json (hunk 8 above)
Comment (wojtek-t) [inline api/swagger-spec/v1.json]: That's a good catch - although this is generated by our scripts.
This is some more generic issue - see e.g. that Get Node is also saying that node is namespaced:
https://github.com/kubernetes/kubernetes/blob/master/api/swagger-spec/v1.json#L2728

So this is unrelated to this PR.
I filed https://github.com/kubernetes/kubernetes/issues/18497 for it.

File: api/swagger-spec/v1.json (hunk 9 above)
Comment (wojtek-t) [inline api/swagger-spec/v1.json]: That's a good catch - although this is generated by our scripts.
This is some more generic issue - see e.g. that Get Node is also saying that node is namespaced:
https://github.com/kubernetes/kubernetes/blob/master/api/swagger-spec/v1.json#L2728

So this is unrelated to this PR.
I filed #18497 for it.

**Comment by wojtek-t** [comment]
@derekwaynecarr - thanks a lot for looking into this PR

>
@wojtek-t - so you are only looking to make the namespace controller use this for Events in this PR? want to make sure I get the initial scope, or if you plan to extend this for other resource types?

Yes - in this PR I'm targeting to change only Events client. This is because @caesarxuchao is working on auto-generating the whole client code, so we will get all other resources for free. (I think other resources are not that urgent - if you think they are, I'm happy to add them, but i would prefer doing it in a separate PR).


I applied all comments - PTAL

**Comment by deads2k** [comment]
> @deads2k - just a heads up, but this PR would allow for DELETE /api/v1/namespace/foo/pods/ to do a DELETE_ALL, which I think is fine for our policy concerns, but wanted to give you a heads up.

It's fine for policy, but will it affect things that are watching for deletion event?  I seem to recall that etcd didn't send deletion events for individual items removed.  Sorry to be lazy; I'd dig through the pull, but its a little large.

**Comment by wojtek-t** [comment]
>
It's fine for policy, but will it affect things that are watching for deletion event? 

@deads2k This problem is solved - watches work fine with it.

**Comment by wojtek-t** [comment]
@derekwaynecarr - PTAL

File: pkg/apiserver/api_installer.go (hunk 7 above)
Comment (derekwaynecarr) [inline pkg/apiserver/api_installer.go]: ok, just wanted to ensure i wasn't crazy that this was not yet actually supported, but i can see how a kind may manage a collection of subresources and we would want delete on that collection.

**Comment by derekwaynecarr** [comment]
@deads2k - yeah, this is basically letting you do a DELETE of the collection to the API server, and then the API server right now makes N calls to etcd so watch behavior is preserved, but we can get around qps throttling to the API server.

@wojtek-t - lgtm, thanks!

**Comment by wojtek-t** [comment]
@derekwaynecarr - thanks a lot for review!

**Comment by wojtek-t** [comment]
@k8s-bot unit test this please

**Comment by wojtek-t** [comment]
@k8s-bot e2e test this please

**Comment by bgrant0607** [comment]
Ref #10217
